### PR TITLE
Enhance subdomain bounding box generator

### LINKED ISSUE #12279: A small enhancement of SubdomainBoundingBox mesh modifier
## Rationale

The enhancement will be restricting the subdomain ID assignment within a bounding box only for a selected set of existing subdomains.

## Description

If no subdomains are specified, assigning subdomain ID will happen for all elements within the bounding box.

## Impact
New feature.

### PR DESCRIPTION
**Pull request by YaqiWang**
To allow reassign subdomain ID for elements only in certain subdomains.
Close #12279.

### CODE REVIEW

**Comment by YaqiWang** [comment]
Reviving my old branch for this because I need this for generating a mesh for a benchmark problem.

File: framework/src/meshgenerators/SubdomainBoundingBoxGenerator.C [hunk 1]
```diff
@@ -38,6 +38,9 @@ SubdomainBoundingBoxGenerator::validParams()
       "block_name", "Subdomain name to set for inside/outside the bounding box (optional)");
   params.addParam<MooseEnum>(
       "location", location, "Control of where the subdomain id is to be set");
+  params.addParam<std::vector<SubdomainName>>(
+      "restricted_subdomains",
+      "Only reset subdomain ID for given subdomains within the bounding box");
```
Comment (zachmprince) [inline framework/src/meshgenerators/SubdomainBoundingBoxGenerator.C:43]: Seems like you can set an empty vector as the default and get rid of `_has_restriction`.

Comment (YaqiWang) [reply framework/src/meshgenerators/SubdomainBoundingBoxGenerator.C:43]: Then this parameter needs to be called excluding subdomains. Note currently if users set the parameter in the input an empty string, this generator will do nothing.

File: test/tests/meshgenerators/subdomain_bounding_box_generator/tests [hunk 2]
```diff
@@ -31,4 +31,16 @@
     mesh_mode = 'REPLICATED'
     recover = false
   [../]
+
+  [./subdomain_bounding_box_generator_restricted_test]
+    type = 'Exodiff'
+    input = 'subdomain_bounding_box_generator_restricted.i'
+    exodiff = 'subdomain_bounding_box_generator_restricted_in.e'
+    cli_args = '--mesh-only'
+    requirement = 'MOOSE shall be able to use the BoundingBox class to assign subdomain IDs for elements with certain subdomain IDs.'
```
Comment (aeslaughter) [inline test/tests/meshgenerators/subdomain_bounding_box_generator/tests:40]:
```suggestion
    requirement = 'The system shall include the ability to assign subdomain identifiers for elements within a bounding box that is restricted to only modify elements with a specific identifier.'
```

**Review by zachmprince — APPROVED** [review]

**Comment by YaqiWang** [comment]
This is ready.

**Pull request merged**